Feature -  5122 - Persist table state

esizer opened the pull request:
🤖 Resolves #5122 

## 👋 Introduction

This persists the state of admin tables in the `URLSearchParams`

## 🕵️ Details

Thile searching, filtering and sorting columns the `URLSearchParams` are updated and used to initialize the state of admin tables.

## ❓ Question

For some reason, this change makes the `PoolTable.tsx` crash because of nested updates. This only occurs on the `PoolTable.tsx` and no other admin tables. I initially thought it was related to the props being passed in but removing those one by one still results in the crash.

Would anyone be able to take a look and assist me with adjusting this to fix that?

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Build admin `npm run production --workspace=admin`
2. Navigate to the different tables
3. Change the state of the table
4. Observe the url update
5. Refresh the page and confirm the initial state matches from before the refresh

tristan-orourke reviewed (commented):
Okay, I haven't found what's different about PoolTable yet, but here are some thoughts from my initial read through of the PR. 

I noticed when testing that whenever a url param is set, that adds entry to the browser history stack. I'm not against that myself, but I believe we collectively decided that changing filters and table state shouldn't affect browser history.

If we do allow it to modify browser history, there's a bug right now where if you use the Back button in the browser, the url param goes to the previous version (as expected), but then the table overwrites it immediately, essentially sending you forward unintentionally.

I would also prefer if we can keep the url cleaner by only adding params that differ from the default. I understand that might be more difficult in the case of User and PoolCandidate filters though.

esizer commented:
> I noticed when testing that whenever a url param is set, that adds entry to the browser history stack. I'm not against that myself, but I believe we collectively decided that changing filters and table state shouldn't affect browser history.

In regards to affecting browser history. This was done explicitly because of the acceptance criteria stating: "then click Back in the Browser...THEN you see the same results as when you left (including the same filters, same page, and same page size)."

However, this does become very annoying because of how reactive our tables are (update every change). I can look into debouncing the search params a bit to see if this helps with how frequently they touch the history. Ideally these only touch browser history once so we get the back button functionality we want without having to hit it like 6 times to get back to the previous page.

> If we do allow it to modify browser history, there's a bug right now where if you use the Back button in the browser, the url param goes to the previous version (as expected), but then the table overwrites it immediately, essentially sending you forward unintentionally.

This might actually be related to the pools table 🤔 . I was getting an infinite update issue and maybe it was

1. Table sets defaults
2. Render
3. URL overrides defaults
4. Re-render
5. Repeat 1-4

> I would also prefer if we can keep the url cleaner by only adding params that differ from the default. I understand that might be more difficult in the case of User and PoolCandidate filters though.

Makes sense to me! I thought I did this but I guess not 😅

tristan-orourke commented:
@esizer I fixed PoolTable with https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/5cb231e56bc83cad25303cd927c011cb18959888!

tristan-orourke commented:
> > If we do allow it to modify browser history, there's a bug right now where if you use the Back button in the browser, the url param goes to the previous version (as expected), but then the table overwrites it immediately, essentially sending you forward unintentionally.

> This might actually be related to the pools table 🤔 . I was getting an infinite update issue and maybe it was

Nope, I've been seeing that issue on all tables, including the User table.

tristan-orourke commented:
> In regards to affecting browser history. This was done explicitly because of the acceptance criteria stating: "then click Back in the Browser...THEN you see the same results as when you left (including the same filters, same page, and same page size)."

Ah, the intention was that clicking Back from a different page would come back to the table, in the state you left it, but if you never left the table, then Back wouldn't bring you to a previous state, it would bring you to wherever you were before the table page. I've updated the acceptance criteria to be more explicit.

tristan-orourke commented on frontend/admin/src/js/components/Table/Table.tsx:
```diff
@@ -170,6 +177,27 @@ function Table<T extends Record<string, unknown>>({
+        }
+      }
+      if (hiddenColumns) {
+        const newHiddenColumns = hiddenColumns.join(",");
```
Should hidden columns also be wrapped in encodeURIComponent? I would have done so in my commit, but I wasn't sure if there was a reason you had done so for sortBy but not hiddenColumns.

esizer replied:
They don't need to. The join is effectively accomplishing the same task. It was used with `sortBy` since it is an object and a little trickier to convert nicely.

esizer commented on frontend/admin/src/js/components/user/UserTable.tsx:
```diff
@@ -167,32 +170,81 @@ export const UserTable = ({ initialFilterInput }: UserTableProps) => {
   const intl = useIntl();
   const paths = useAdminRoutes();
   const { pathname } = useLocation();
+  const [searchParams, setSearchParams] = useSearchParams();
+  const {
+    currentPage: initialPage,
+    pageSize: initialPageSize,
+    hiddenColumnIds: initialHiddenColumns,
+    sortBy: initialSortBy,
+    searchState: initialSearchState,
+  } = getCommonTableParams(searchParams);
```
Agreed 👍 . I did that to the client side table as a cleanup then ran into the `PoolTable` issues and kinda just stopped working through this PR 😜 

I can update this to be similar to the other hook.

esizer commented:
> @esizer I fixed PoolTable with [5cb231e](https://github.com/GCTC-NTGC/gc-digital-talent/commit/5cb231e56bc83cad25303cd927c011cb18959888)! It was hiddenColumns that was really the real problem.

Does not seem to fix for me 😢

esizer commented on frontend/admin/src/js/components/Table/Table.tsx:
```diff
@@ -140,6 +142,9 @@ function Table<T extends Record<string, unknown>>({
   hiddenCols = [],
   initialSortBy = [],
 }: TableProps<T>): ReactElement {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const initialState = useInitialTableState(searchParams);
```
We need access to the state returned by `react-table` to store the new state in the `useEffect` and since the `initialState` is used in setting up the table, it needs to be defined first so there is no way to do this in a clean way unless we move the `useTable` hook there as well.

tristan-orourke replied:
We could still wrap useSearchParams in `useInitialTableState()`, and expose setSearchParams as setInitialState: `const [initialState, setInitialState] = useInitialTableState();`

esizer replied:
This is another case of me being stumped on how to get this to work and not result in an infinite render loop. The reason to decoupling all of this was to ensure `initialState` was very stable and unchanging since any change to the initial state causes a re-render which ends up in an infinite re-render. 😕

tristan-orourke commented:
> > @esizer I fixed PoolTable with [5cb231e](https://github.com/GCTC-NTGC/gc-digital-talent/commit/5cb231e56bc83cad25303cd927c011cb18959888)! It was hiddenColumns that was really the real problem.
> 
> Does not seem to fix for me 😢

Oh no! I'll take another look then.

esizer commented on frontend/admin/src/js/components/Table/Table.tsx:
```diff
@@ -170,6 +177,22 @@ function Table<T extends Record<string, unknown>>({
+
+      return previous;
+    });
+  }, [pageSize, pageIndex, hiddenColumns, sortBy, setSearchParams]);
```
This should be fixed by [delete default params](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/6991c9ae3ad2168460eaf2e502abfbb3ad226c50)

esizer commented on frontend/admin/src/js/components/apiManagedTable/basicTableHelpers.tsx:
```diff
@@ -186,3 +191,74 @@ export function useDebounce<T>(value: T, delay: number): T {
+) {
+  if (pageSize) {
+    currentParams.set("pageSize", `${pageSize}`);
+  }
```
Same as previous, should be fixed by [delete default params](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/6991c9ae3ad2168460eaf2e502abfbb3ad226c50)

esizer commented:
> Nope, I've been seeing that issue on all tables, including the User table.

> Ah, the intention was that clicking Back from a different page would come back to the table, in the state you left it, but if you never left the table, then Back wouldn't bring you to a previous state, it would bring you to wherever you were before the table page. I've updated the acceptance criteria to be more explicit.

Updated to the the `replace: true` option in search params so it does not add to the stack. This still preserves the requested functionality of restoring state with the back button after leaving the page. Done in two separare commits for api and client tables:

- [fix bug with back button overriding url](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/bc606f63f56176057703de50f7cac0d5391bd59d)
- [fix history stack for client table](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/fda2bdb519fca1076d2d84290a8c975d76545661)

esizer commented on frontend/admin/src/js/components/poolCandidate/PoolCandidatesTable.tsx:
```diff
@@ -275,6 +300,42 @@ const PoolCandidatesTable: React.FC<{
+      },
+    );
+
+    setSearchParams(newSearchParams);
```
Fixed with [replace history on pool candidate table](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/66aa727741a081697e9850a089229b71770078c7)

esizer commented on frontend/admin/src/js/components/Table/Table.tsx:
```diff
@@ -170,6 +182,53 @@ function Table<T extends Record<string, unknown>>({
+          previous.delete("pageIndex");
+        }
+
+        if (sortBy) {
```
Fixed with [update table state comparisons](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/fb5fbe547cd85a3f6e2593a8532400bd917df4f9)

esizer commented on frontend/admin/src/js/components/Table/Table.tsx:
```diff
@@ -170,6 +182,53 @@ function Table<T extends Record<string, unknown>>({
+          previous.delete("sortBy");
+        }
+
+        if (hiddenColumns && hiddenColumns.length) {
```
Fixed as part of [update table state comparisons](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/fb5fbe547cd85a3f6e2593a8532400bd917df4f9)

esizer commented:
> I still think it would be nice to combine useSearchParams, getCommonTableParams, setCommonTableParams, and setSearchParams for other filters, into a single hook. It would make new tables easier - wouldn't have to worry about forgetting setSearchParams(..., {replace: true}) anywhere, for example. I can take a stab at it, if you want.

I agree, do you have something in mind? My only concern is we need it in two different shapes to interact with the custom filters easily. The getters and setters are `URLSearchParams` which can only use strings, however when we actually use that info it needs to typed properly using the `CommonTableParams` interface. I guess we could return three items rather than the usual tuple 🤔 

```tsx
interface CommonTableParams<T> {
  currentPage?: number;
  pageSize?: number;
  hiddenColumnIds?: Array<string>;
  sortBy?: SortingRule<T>;
  searchState?: SearchState;
}

type UseCommonTableParams<T> = () => {
  params: URLSearchParams,
  state: CommonTableParams<T>,
  setParams: SetURLSearchParams
};
```

tristan-orourke commented:
I've been thinking something like this:

```
interface CommonTableParams<T> {
  currentPage?: number;
  pageSize?: number;
  hiddenColumnIds?: Array<string>;
  sortBy?: SortingRule<T>;
  searchState?: SearchState;
}

interface TableState<T, F> extends CommonTableParams<T> {
  filters: F,
}

// Alternatively, if we like the look more...
// interface TableState<T, F> {
//   commonParams:  CommonTableParams<T>,
//   filters: F,
// }
   

type UseCachedTableState<T, F = never> = (
  defaultState: Partial<TableState<T,F>> // newState will only be stored in the cache if it differs from defaultState.
) => [
  state: TableState,
  setState: (newState: TableState<T, F>) => void,
];

esizer commented:
@tristan-orourke Okay, so this is only working for `UserTable.tsx` but I have a working implementation of your suggestion. Wanted you to take a look before I take it a step further. I think we could even simplify this further and not need to separate `useState` for each item in the table itself.

[move table state to separate hook](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/6019bc5f1920fa36554a62dda1e7b5f0171827f6)

tristan-orourke commented:
https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/6019bc5f1920fa36554a62dda1e7b5f0171827f6 looks great! Very smart, how you separated the setter functions out into type-based helper functions.

tristan-orourke commented on frontend/admin/src/js/components/apiManagedTable/useTableState.ts:
```diff
@@ -0,0 +1,210 @@
+) => {
+  const prevValue = params.get(key);
+  if (!prevValue || !isEqual(prevValue.split(","), newValue)) {
+    if (newValue && newValue !== defaultValue) {
```
This should also use deep comparison, no?
```suggestion
    if (newValue && !isEqual(newValue, defaultValue)) {
```

esizer replied:
Yes! Thanks for catching that 😄

esizer commented:
Lol another interesting one I cannot figure out @tristan-orourke  😆 . For some reason, without making any changes to the client table, it started throwing the update depth warning on all tables now. Removing some of the items from the `useEffect` fixes it but breaks the exhaustive deps eslint rule. It is specifically the `initialSortBy` that is causing it but that shouldnt even be changing since it is only defined on the `PoolTable` 🤔 

https://github.com/GCTC-NTGC/gc-digital-talent/blob/3566ab0ecb2a7445dec4a33ca876f777403d5d80/frontend/admin/src/js/components/Table/Table.tsx#L237

tristan-orourke commented:
Found it. The default values for props were defining a new empty list every time the Table component rendered, and that re-triggered the useEffect.
Fixed in https://github.com/GCTC-NTGC/gc-digital-talent/pull/5223/commits/6fcd72ed34eea5224c3a9526e8ce85776d354d89

tristan-orourke reviewed (approved):
Looks great! Thanks for working with me on those hooks.

esizer commented:
> Looks great! Thanks for working with me on those hooks.

Thanks for the help! Working through the hooks was definitely worth it 👍
